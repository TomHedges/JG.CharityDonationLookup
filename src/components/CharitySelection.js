import React, { Component } from "react";
import { hot } from "react-hot-loader";
import PropTypes from "prop-types";

const CharitySelection = props => {
  const listItems = props.charity_slugs.map(slug => (
    <span key={slug} onClick={props.handleClick}>
      {slug}
    </span>
  ));

  const dropdown_class =
    props.show_list === true
      ? "dropdown_content_shown"
      : "dropdown_content_hidden";

  return (
    <div className="charity_selection">
      <form onSubmit={props.handleSubmit}>
        <label htmlFor="charity_id">
          Please enter Charity ID and select charity from dropdown list:
        </label>
        <div className="dropdown_wrapper">
          <input
            id="charity_id"
            type="text"
            className="charity_id_selector"
            value={props.charity_ID}
            onChange={props.handleChange}
          />
          <div className={dropdown_class}>{listItems}</div>
        </div>
      </form>
    </div>
  );
};

CharitySelection.propTypes = {
  handleChange: PropTypes.func.isRequired,
  handleClick: PropTypes.func.isRequired,
  handleSubmit: PropTypes.func.isRequired,
  charity_slugs: PropTypes.array.isRequired,
  charity_ID: PropTypes.string.isRequired,
  charity_selected: PropTypes.bool.isRequired,
  show_list: PropTypes.bool.isRequired
};

export default hot(module)(CharitySelection);
