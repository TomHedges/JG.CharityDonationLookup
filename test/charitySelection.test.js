import CharitySelection from "../src/components/CharitySelection";
import { handleSubmit } from "../src/scripts/event_handlers.js/index.js";
import renderer from "react-test-renderer";

/**
 * Initialize common properties to be passed
 * @param {*} props properies to be override
 */

const TEST_CHARITY_SLUG = "Tom's Great Charity (12345)";

function createTestProps(props) {
  return {
    handleChange: () => {
      target: {
        value: "15678";
      }
    },
    handleClick: () => {},
    handleSubmit: handleSubmit,
    charity_ID: "123",
    charity_slug: TEST_CHARITY_SLUG,
    show_list: true,
    ...props
  };
}

let wrapper, props;

beforeEach(() => {
  props = createTestProps();
  wrapper = shallow(<CharitySelection {...props} />);
});

describe("<CharitySelection /> rendering", () => {
  test("Should render <CharitySelection /> component", () => {
    expect(wrapper.exists()).toBe(true);
  });
});

describe("<CharitySelection /> interactions", () => {
  /*
  test("Record change in input value when text is entered", () => {
    //console.log(wrapper.find("input").props());
    wrapper.find("input").simulate("change", {
      target: { value: "15678" }
    });
    //wrapper = shallow(<CharitySelection {...props} charity_ID={"15678"} />);
    //console.log(wrapper.find("input").props());
    expect(wrapper.find("input").props().value).toEqual("15678");
  });
  */

  test("Should cancel form submission", () => {
    let prevented = false;
    wrapper.find("form").simulate("submit", {
      preventDefault: () => {
        prevented = true;
      }
    });
    expect(prevented).toBe(true);
  });
});

describe("<CharitySelection /> display changes based on state", () => {
  test("Should display charity_slug text", () => {
    expect(wrapper.find("div.dropdown_content_shown")).toHaveLength(1);
  });

  test("Should not display charity_slug text when show_list is false", () => {
    wrapper = shallow(<CharitySelection {...props} show_list={false} />);
    expect(wrapper.find("div.dropdown_content_shown")).toHaveLength(0);
  });

  test("Should not display charity_slug dropdown when message is blank", () => {
    wrapper = shallow(<CharitySelection {...props} charity_slug={""} />);
    expect(wrapper.find("div.dropdown_content_shown span")).toHaveLength(0);
  });

  /*
  test("Should not display charity_slug dropdown when message is null", () => {
    wrapper = shallow(<CharitySelection {...props} charity_slug={null} />);
    expect(wrapper.find("div.dropdown_content_shown span")).toHaveLength(0);
  });
  */

  let tree = renderer;

  test("Snapshot should contain charity_slug text in span, within div class 'dropdown_content_shown'", () => {
    tree = renderer.create(wrapper).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test("Snapshot should contain charity_slug text in span, within div class 'dropdown_content_hidden'", () => {
    wrapper = shallow(<CharitySelection {...props} show_list={false} />);
    tree = renderer.create(wrapper).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test("Snapshot should not contain charity_slug text in span, within div class 'dropdown_content_hidden'", () => {
    wrapper = shallow(<CharitySelection {...props} show_list={false} />);
    tree = renderer.create(wrapper).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
