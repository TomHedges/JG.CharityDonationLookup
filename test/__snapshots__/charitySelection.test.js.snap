// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CharitySelection /> display changes based on state Snapshot should contain charity_slug text in span, within div class 'dropdown_content_hidden' 1`] = `
<div
  className="charity_selection"
>
  <form
    onSubmit={[Function]}
  >
    <label
      htmlFor="charity_id"
    >
      Please enter Charity ID and select charity from dropdown list:
    </label>
    <div
      className="dropdown_wrapper"
    >
      <input
        className="charity_id_selector"
        id="charity_id"
        onChange={[Function]}
        type="text"
        value="123"
      />
      <div
        className="dropdown_content_hidden"
      >
        <span
          className="charity_result"
          onClick={[Function]}
        >
          Tom's Great Charity (12345)
        </span>
      </div>
    </div>
  </form>
</div>
`;

exports[`<CharitySelection /> display changes based on state Snapshot should contain charity_slug text in span, within div class 'dropdown_content_shown' 1`] = `
<div
  className="charity_selection"
>
  <form
    onSubmit={[Function]}
  >
    <label
      htmlFor="charity_id"
    >
      Please enter Charity ID and select charity from dropdown list:
    </label>
    <div
      className="dropdown_wrapper"
    >
      <input
        className="charity_id_selector"
        id="charity_id"
        onChange={[Function]}
        type="text"
        value="123"
      />
      <div
        className="dropdown_content_shown"
      >
        <span
          className="charity_result"
          onClick={[Function]}
        >
          Tom's Great Charity (12345)
        </span>
      </div>
    </div>
  </form>
</div>
`;

exports[`<CharitySelection /> display changes based on state Snapshot should not contain charity_slug text in span, within div class 'dropdown_content_hidden' 1`] = `
<div
  className="charity_selection"
>
  <form
    onSubmit={[Function]}
  >
    <label
      htmlFor="charity_id"
    >
      Please enter Charity ID and select charity from dropdown list:
    </label>
    <div
      className="dropdown_wrapper"
    >
      <input
        className="charity_id_selector"
        id="charity_id"
        onChange={[Function]}
        type="text"
        value="123"
      />
      <div
        className="dropdown_content_hidden"
      >
        <span
          className="charity_result"
          onClick={[Function]}
        >
          Tom's Great Charity (12345)
        </span>
      </div>
    </div>
  </form>
</div>
`;
